source_test.go

```diff
@@ -0,0 +1,638 @@
+package dalec
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"path/filepath"
+	"reflect"
+	"slices"
+	"strings"
+	"testing"
+
+	"github.com/moby/buildkit/client/llb"
+	"github.com/moby/buildkit/exporter/containerimage/image"
+	"github.com/moby/buildkit/frontend/dockerfile/dockerfile2llb"
+	"github.com/moby/buildkit/solver/pb"
+	"github.com/opencontainers/go-digest"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+func TestSourceGitSSH(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+
+	// buildkit's llb.Git currently directly runs an ssh keyscan (connects to the specified host and port to get the host key)
+	// This is not ideal for our test setup here because we need to use a real server.
+	// Thankfully when there is an error it is ignored so we don't actually need to setup a full SSH server here.
+	addr := stubListener(t)
+
+	src := Source{
+		Git: &SourceGit{
+			URL:    fmt.Sprintf("user@%s:test.git", addr),
+			Commit: t.Name(),
+		},
+	}
+
+	ops := getSourceOp(ctx, t, src)
+	checkGitOp(t, ops, &src)
+
+	t.Run("with subdir", func(t *testing.T) {
```

Comment:
It looks like the subpath/include-exclude setup is duplicated across different tests. Maybe you could use an options type pattern to set a subdir and include/exclude on the source?

Reply:
I feel like this isn't really enough logic to make it worth abstracting.